File: dashboard/src/lib/windows.ts
import type { Aggregates, WeeklyEntry, Windows } from "./aggregates.gen";

export type AnyWindow = Windows[number];

// "YYYY-Www" sorts lexicographically (zero-padded, year first), so plain
// string comparison is correct week ordering throughout.

/** Default filter value: all_time, the widest view. Falls back to the semester
 *  with the latest coverage if no all_time window is published. */
export function defaultWindowId(windows: Windows): string {
  const semesters = windows
    .filter((w) => w.kind === "semester")
    .sort((a, b) => (a.coverage.through < b.coverage.through ? 1 : -1));
  return windows.find((w) => w.kind === "all_time")?.id ?? semesters[0]?.id ?? windows[0]?.id ?? "";
}

export function findWindow(windows: Windows, id: string | null): AnyWindow | undefined {
  return windows.find((w) => w.id === id);
}

/** A semester whose published coverage stops short of its final member week is
 *  still accumulating data. (Contract §6.1 — no client-side date math needed.)
 *
 *  Only semesters can be in progress: a slice's weeks *are* its coverage, so it is always
 *  complete by construction, and its own label ("Last available week") is true either way.
 *  The marker therefore lands on the semester option and nowhere else. */
export function isInProgress(win: AnyWindow): boolean {
  if (win.kind !== "semester") return false;
  return win.weeks.length > 0 && win.coverage.through < win.weeks[win.weeks.length - 1];
}

/** Display selection, not re-aggregation (invariant 4): weekly series are
 *  published over the full range; a window filter just narrows what is shown. */
export function sliceToWindow(series: WeeklyEntry[], win: AnyWindow): WeeklyEntry[] {
  return series.filter((e) => e.week >= win.coverage.from && e.week <= win.coverage.through);
}

/** Does this window hold exactly one ISO week of data? Some measures are defined away by
 *  that — weeks active per student is 1 for everyone, whatever the data says.
 *
 *  Read off `coverage`, never `weeks`: a semester's `weeks` is its full Thursday-rule
 *  membership (17–18 entries from the day it opens), so the first week of a new term is a
 *  semester with one covered week — the same single week its `.last1` slice holds, and the
 *  default landing state at that moment. Asking `weeks.length` there answers "how long is
 *  the term" when the question was "how much data is in this window". Coverage is clipped
 *  to covered weeks for every kind and the axis is dense, so its endpoints are exact. */
export function isSingleWeek(win: AnyWindow): boolean {
  return win.coverage.from === win.coverage.through;
}

/** Is this window a slice of a semester (D-56)? Narrows the union for `parent_window_id`. */
export function parentWindowId(win: AnyWindow): string | null {
  return win.kind === "semester_slice" ? win.parent_window_id : null;
}

/**
 * Picker groups, in display order: semesters (newest first), the anchor semester's slices,
 * then all-time.
 *
 * Three fixed groups, not one per semester (D-57). Only the anchor is sliced, so nesting
 * would produce two headings holding a single option each and repeat the same three words
 * under the third. Slices are filtered by `kind` rather than by parent: whichever semester
 * they belong to, they are "Recent", and their own labels name it ("Previous 4 weeks ·
 * SS 2026"). Registry order is display order within a group — the pipeline emits the wider
 * slice first, and that is its statement to make, not ours to re-sort.
 */
export function groupedWindows(doc: Aggregates): { label: string; windows: AnyWindow[] }[] {
  const semesters = doc.windows
    .filter((w) => w.kind === "semester")
    .sort((a, b) => (a.coverage.through < b.coverage.through ? 1 : -1));
  const recent = doc.windows.filter((w) => w.kind === "semester_slice");
  const allTime = doc.windows.filter((w) => w.kind === "all_time");
  return [
    { label: "Semesters", windows: semesters },
    { label: "Recent", windows: recent },
    { label: "Everything", windows: allTime },
  ].filter((g) => g.windows.length > 0);
}

File: pipeline/statsboteval_pipeline/windows.py
"""Windows registry (contract §6.1): all_time, semesters, semester slices.

A week belongs to the semester containing its Thursday. SS = Mar 1-Jun 30,
WS = Oct 1-Jan 31 (following year); Feb and Jul-Sep weeks are break weeks and
belong only to all_time. Semester `weeks` list the full calendar membership;
`coverage` is clipped to the axis (the data range).

The **anchor** semester — the one with the latest coverage — also publishes two slices of
its closing stretch (D-56, narrowed to the anchor alone at D-57): the last covered week and
the last up-to-`SLICE_WEEKS` covered weeks. These replace the old axis-anchored
`trailing_4`, which advanced with extraction whether or not anyone was in class — across a
break it drifted into weeks holding almost nothing.
"""

from __future__ import annotations

from collections.abc import Sequence
from datetime import date, timedelta
from typing import NamedTuple

from .contract import (
    AllTimeWindow,
    Coverage,
    SemesterSliceWindow,
    SemesterWindow,
    Window,
    date_to_week,
    week_monday,
)

SLICE_WEEKS = 4


class _Semester(NamedTuple):
    id: str
    label: str
    short: str
    start_date: date
    end_date: date


def _semester_of(week: str) -> _Semester | None:
    """The semester owning this week (Thursday rule), or None for break weeks."""
    thursday = week_monday(week) + timedelta(days=3)
    y, m = thursday.year, thursday.month
    if 3 <= m <= 6:
        return _Semester(f"{y}S", f"Summer semester {y}", f"SS {y}", date(y, 3, 1), date(y, 6, 30))
    if m >= 10:
        return _winter(y)
    if m == 1:
        return _winter(y - 1)
    return None


def _winter(y: int) -> _Semester:
    """Both winter branches build the same tuple; two spellings of "2025/26" would drift."""
    spanning = f"{y}/{(y + 1) % 100:02d}"
    return _Semester(f"{y}W", f"Winter semester {spanning}", f"WS {spanning}", date(y, 10, 1), date(y + 1, 1, 31))


def _semester_weeks(start: date, end: date) -> list[str]:
    """Full membership: every week whose Thursday lies within [start, end]."""
    out: list[str] = []
    monday = week_monday(date_to_week(start))
    while monday <= end:
        if start <= monday + timedelta(days=3) <= end:
            out.append(date_to_week(monday))
        monday += timedelta(days=7)
    return out


def _slices(semester: SemesterWindow, covered: Sequence[str], short: str) -> list[SemesterSliceWindow]:
    """The anchor semester's closing stretch, as one-week and up-to-SLICE_WEEKS windows.

    Labels are state-free (D-57). D-56 branched them on whether the term was still running
    ("Latest" vs "Final"), because the same week-set answers "how is it going" for a live
    term and "how did it close" for a finished one. "Previous 4 weeks" and "Last available
    week" are true readings of both, so the branch was buying a distinction the words
    already carry — and the picker shows "(in progress)" on the semester itself anyway.

    The multi-week label states the count it *actually* holds. Early in a term there are
    fewer than SLICE_WEEKS weeks to take, and saying "4 weeks" over two of them is exactly
    the failure that made the old trailing window need renaming to "Last Avl. 4 weeks"
    instead of fixing.
    """
    out: list[SemesterSliceWindow] = []

    def slice_for(weeks: Sequence[str], suffix: str, short_label: str) -> SemesterSliceWindow:
        return SemesterSliceWindow(
            kind="semester_slice",
            id=f"{semester.id}.{suffix}",
            # "Last available week · SS 2026": self-contained, because the picker no longer
            # nests these under a heading that names the semester (D-57) and because the
            # page prints a window name into sentences — card captions, empty states, gaps.
            label=f"{short_label} · {short}",
            short_label=short_label,
            parent_window_id=semester.id,
            weeks=list(weeks),
            # Indexed against full membership, never `covered` — see contract._check_windows.
            semester_weeks=[
                semester.weeks.index(weeks[0]) + 1,
                semester.weeks.index(weeks[-1]) + 1,
            ],
            coverage=Coverage(from_=weeks[0], through=weeks[-1]),
        )

    # Widest first: the picker renders registry order, and "Recent" reads wider -> narrower.
    if len(covered) >= 2:
        multi = covered[-SLICE_WEEKS:]
        out.append(slice_for(multi, f"last{SLICE_WEEKS}", f"Previous {len(multi)} weeks"))
    # At one covered week the multi-week slice would be a second name for this same week,
    # so it is omitted above rather than published as a duplicate under a different id.
    out.append(slice_for(covered[-1:], "last1", "Last available week"))
    return out


def build_windows(axis: Sequence[str]) -> list[Window]:
    """Registry for a dense, chronological, complete-week axis (aggregate's invariant)."""
    if not axis:
        raise ValueError("axis is empty; no windows to build")

    windows: list[Window] = [
        AllTimeWindow(
            kind="all_time",
            id="all_time",
            label="All time",
            coverage=Coverage(from_=axis[0], through=axis[-1]),
        )
    ]

    seen: set[str] = set()
    # The anchor is simply the last semester appended: the axis is chronological and `seen`
    # gates first appearance, so semesters accumulate in order and whichever is assigned
    # last is the one with the latest coverage. Same rule the dashboard's `defaultWindowId`
    # applies client-side — stated once here, in Python, where the calendar knowledge lives.
    anchor: tuple[SemesterWindow, list[str], str] | None = None
    for week in axis:
        sem = _semester_of(week)
        if sem is None or sem.id in seen:
            continue
        seen.add(sem.id)
        weeks = _semester_weeks(sem.start_date, sem.end_date)
        covered = [w for w in axis if w in set(weeks)]
        semester = SemesterWindow(
            kind="semester",
            id=sem.id,
            label=sem.label,
            start_date=sem.start_date,
            end_date=sem.end_date,
            weeks=weeks,
            coverage=Coverage(from_=covered[0], through=covered[-1]),
        )
        windows.append(semester)
        anchor = (semester, covered, sem.short)

    # Only the anchor is sliced (D-57). D-56 sliced every semester so closing stretches
    # could be compared across terms; that comparison turned out not to be wanted, and the
    # fan-out cost four unselectable windows and ~205 KB in every document.
    if anchor is not None:
        windows.extend(_slices(*anchor))
    return windows
